### DISCUSSION TITLE
Unused functions

### DISCUSSION THREAD

**Discussion by roccoblues** [neomutt/neomutt#4118] [Development]
I ran [xunused](https://github.com/mgehre/xunused) over our codebase and found a couple of unused functions. Usage only in `test/` counts as unused.

Some of them probably make sense to keep, for example `buf_upper` is unused but the opposite `buf_lower` is used. But some might be fine to remove. Here's the list, let me know how you want to proceed.

```
color/parse_ansi.c:75: warning: Function 'ansi_skip_sequence' is unused
mutt/buffer.c:715: warning: Function 'buf_upper' is unused
config/set.c:1083: warning: Function 'cs_str_delete' is unused
config/set.c:885: warning: Function 'cs_str_native_get' is unused
config/set.c:706: warning: Function 'cs_str_string_get' is unused
config/set.c:1044: warning: Function 'cs_str_string_minus_equals' is unused
config/set.c:961: warning: Function 'cs_str_string_plus_equals' is unused
config/subset.c:524: warning: Function 'cs_subset_str_delete' is unused
config/subset.c:266: warning: Function 'cs_subset_str_native_get' is unused
config/subset.c:339: warning: Function 'cs_subset_str_reset' is unused
config/subset.c:483: warning: Function 'cs_subset_str_string_minus_equals' is unused
config/subset.c:445: warning: Function 'cs_subset_str_string_plus_equals' is unused
editor/enter.c:413: warning: Function 'editor_buffer_get_cursor' is unused
editor/enter.c:400: warning: Function 'editor_buffer_get_lastchar' is unused
editor/enter.c:448: warning: Function 'editor_buffer_set' is unused
editor/enter.c:430: warning: Function 'editor_buffer_set_cursor' is unused
mutt/logging.c:496: warning: Function 'log_disp_null' is unused
mutt/file.c:1110: warning: Function 'mutt_file_chmod' is unused
mutt/file.c:1191: warning: Function 'mutt_file_chmod_rm' is unused
key/get.c:177: warning: Function 'mutt_flush_unget_to_endcond' is unused
gui/curs_lib.c:170: warning: Function 'mutt_perror_debug' is unused
mutt/random.c:121: warning: Function 'mutt_rand32' is unused
conn/socket.c:342: warning: Function 'mutt_socket_buffer_readln_d' is unused
mutt/string.c:734: warning: Function 'mutt_str_lws_rlen' is unused
mutt/string.c:825: warning: Function 'mutt_str_next_word' is unused
progress/progress.c:191: warning: Function 'progress_set_size' is unused
config/quad.c:244: warning: Function 'quad_str_toggle' is unused
mutt/slist.c:58: warning: Function 'slist_add_list' is unused
mutt/slist.c:141: warning: Function 'slist_empty' is unused
```

**Comment by flatcap** [comment]
> I ran [xunused](https://github.com/mgehre/xunused) over our codebase

Ooh!  That's nice!

Would you mind adding a page to our [analysis section](https://neomutt.org/dev/analysis)?
A few lines linking the project and describing how you ran it.
It's just markdown (the page links are auto-magic).
You can cut'n'paste one of the other files, e.g. [iwyu.md](https://github.com/neomutt/neomutt.github.io/blob/main/_dev/analysis/iwyu.md)

**Reply by roccoblues** [reply]
done: https://neomutt.org/dev/analysis/xunused

**Comment by flatcap** [comment]
I had a look through the functions.
From their names alone, my first thought was "keep" for most of them.
Mainly for symmetry with other functions :-)

But the longer I think about it the more I'm shifting to, "kill 'em all".

My only hesitation is the fact that we're deleting code that _might_ get used and _probably_ works.
Perhaps we should leave a comment in the code, `// Removed: buf_upper()` in case anyone comes looking.

**Comment by flatcap** [comment]
OK, here are my thoughts.
Take with a pinch of salt :-)
Thanks.

### Kill, no comment

Obsolete code:

```
color/parse_ansi.c:75:   ansi_skip_sequence
key/get.c:177:           mutt_flush_unget_to_endcond
mutt/random.c:121:       mutt_rand32
mutt/string.c:734:       mutt_str_lws_rlen
mutt/string.c:825:       mutt_str_next_word
```

### Kill, leave comment

These functions _could_ be useful.
Leave a comment in the `@page` or the header file.

```
config/quad.c:244:       quad_str_toggle
config/set.c:706:        cs_str_string_get
config/set.c:885:        cs_str_native_get
config/set.c:961:        cs_str_string_plus_equals
config/set.c:1044:       cs_str_string_minus_equals
config/set.c:1083:       cs_str_delete
config/subset.c:266:     cs_subset_str_native_get
config/subset.c:339:     cs_subset_str_reset
config/subset.c:445:     cs_subset_str_string_plus_equals
config/subset.c:483:     cs_subset_str_string_minus_equals
config/subset.c:524:     cs_subset_str_delete
gui/curs_lib.c:170:      mutt_perror_debug
mutt/buffer.c:715:       buf_upper
mutt/file.c:1110:        mutt_file_chmod
mutt/file.c:1191:        mutt_file_chmod_rm
mutt/slist.c:58:         slist_add_list
mutt/slist.c:141:        slist_empty
```

### Keep

This function is used by the tests and the fuzzer, but it's trivial.

```
mutt/logging.c:496:      log_disp_null
```

Copy it into both:
- `fuzz/address.c`
- `test/common.c`

---

This is a new function which is used by as-yet-unmerged maildir code.

```
progress/progress.c:191: progress_set_size
```

---

These functions are used only by the tests.

```
editor/enter.c:400:      editor_buffer_get_lastchar
editor/enter.c:413:      editor_buffer_get_cursor
editor/enter.c:430:      editor_buffer_set_cursor
editor/enter.c:448:      editor_buffer_set
```

Delete the tests that actually test these functions,
and move the functions into a new file: `test/editor/common.c`

---

This function is used by a macro: `#define mutt_socket_buffer_readln`

```
conn/socket.c:342:       mutt_socket_buffer_readln_d
```

**Reply by roccoblues** [reply]
> ### Kill, no comment
> 
> Obsolete code:
> 
> ```
> color/parse_ansi.c:75:   ansi_skip_sequence
> key/get.c:177:           mutt_flush_unget_to_endcond
> mutt/random.c:121:       mutt_rand32
> mutt/string.c:734:       mutt_str_lws_rlen
> mutt/string.c:825:       mutt_str_next_word
> ```

done here: https://github.com/neomutt/neomutt/pull/4120

**Comment by roccoblues** [comment]
- [x] add a weekly Github action that runs [xunused](https://github.com/mgehre/xunused)

**Comment by roccoblues** [comment]
> These functions are used only by the tests.
> 
> ```
> editor/enter.c:400:      editor_buffer_get_lastchar
> editor/enter.c:413:      editor_buffer_get_cursor
> editor/enter.c:430:      editor_buffer_set_cursor
> editor/enter.c:448:      editor_buffer_set
> ```
> 
> Delete the tests that actually test these functions, and move the functions into a new file: `test/editor/common.c`

done here: https://github.com/neomutt/neomutt/pull/4131

**Comment by roccoblues** [comment]
> This function is used by the tests and the fuzzer, but it's trivial.
> 
> ```
> mutt/logging.c:496:      log_disp_null
> ```
> 
> Copy it into both:
> 
>     * `fuzz/address.c`
>     * `test/common.c`

done here: https://github.com/neomutt/neomutt/pull/4133

**Comment by roccoblues** [comment]
All done. First successful "Unused Functions" run is here: https://github.com/neomutt/neomutt/actions/runs/7252509759/job/19757136629

**Reply by flatcap** [reply]
Excellent!
